sql: implement GLOBAL/REGIONAL BY TABLE -> REGIONAL BY ROW change

otan opened the pull request:
This commit adds the ability to change the locality of a GLOBAL or
REGIONAL BY TABLE table to REGIONAL BY ROW, provided an existing column
of the appropriate type already exists.

We note that changing to REGIONAL BY ROW changes the partitioning of all
indexes of the table to have a prefix with the crdb_internal_type'd
column table. This logic already exists in ALTER PRIMARY KEY, so our
only job is to include to implicit partitioning columns and
modify the ALTER PRIMARY KEY schema changer to update zone configs as
appropriate and change the locality config when the mutation is
complete. We add an extra field "LocalityConfigSwap" on the
PrimaryKeyMutation to facilitate this transformation.

Refs: #59632
Resolves #58341 

Release note (sql change): ALTER TABLE ... SET LOCALITY now works to
REGIONAL BY ROW AS col.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59624)
<!-- Reviewable:end -->

otan committed:
sql: implement GLOBAL/REGIONAL BY TABLE -> REGIONAL BY ROW change

This commit adds the ability to change the locality of a GLOBAL or
REGIONAL BY TABLE table to REGIONAL BY ROW, provided an existing column
of the appropriate type already exists.

We note that changing to REGIONAL BY ROW changes the partitioning of all
indexes of the table to have a prefix with the crdb_internal_type'd
column. This logic already exists in ALTER PRIMARY KEY, so our
only job is to include the implicit partitioning columns and
modify the ALTER PRIMARY KEY schema changer to update zone configs as
appropriate and change the locality config when the mutation is
complete. We add an extra field "LocalityConfigSwap" on the
PrimaryKeyMutation to facilitate this transformation.

Release note (sql change): ALTER TABLE ... SET LOCALITY now works to
REGIONAL BY ROW AS col.

ajstorm reviewed (changes requested):
Nice change!  This is really coming together to form a pretty elegant solution with a lot of code reuse (and not a lot of arm-twisting or hackery).  Very impressive.

I also like the _very_ detailed commit record.  Sets the tone for what's to come.  Nice!

>We note that changing to REGIONAL BY ROW changes the partitioning of all
indexes of the table to have a prefix with the crdb_internal_type'd
column table. 

Nit: "table" at the end seems unnecessary.

> so our only job is to include to implicit partitioning columns 

Nit: second "to" seems out of place.  Should it be "the"?


The second commit looks OK to me, but I feel very out of my element reviewing it.  I'd feel more comfortable if @arulajmani could also have a quick look at it.

> Reviewed 11 of 11 files at r1, 1 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-:-MSXmtDUAVFhOnpcjK_8:b-g6ghqe)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, and @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 723 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSXqSRh4kM9Sz1U4Qxe:-MSXqSRh4kM9Sz1U4Qxf:b-pdqbbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L723)):*
> ```
>                                    lease_preferences = '[[+region=ca-central-1]]'
> 
> # Drop the table and recreate it with columns appropriate for REGIONAL BY ROW transformations.
> ```

Nit: This test was meant to test going to RBR (from RBT <no region>) without the column there.  The new test you've added is great, but I want to make sure that we don't lose the old test in the process.  Can you add back an unimplemented test where we first try to move it to RBR without specifying the AS column?
___
*[pkg/ccl/partitionccl/regional_by_row_test.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSYApGO4xfpBjEFK0eg:-MSYApGO4xfpBjEFK0eh:b-d7h4ip) ([raw file](https://github.com/cockroachdb/cockroach/blob/298e651a6a1318c7a3727e88b7d8961a7f06fb0e/pkg/ccl/partitionccl/regional_by_row_test.go#L56)):*
> ```Go
> 	params, _ := tests.CreateTestServerParams()
> 	params.Locality.Tiers = []roachpb.Tier{
> 		{Key: "region", Value: "ajstorm-1"},
> ```

Nit: checking if I'm paying attention?  :joy: 
___
*[pkg/sql/alter_primary_key.go, line 232 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSXx87f9l_jy8XTOq6N:-MSXx87g5jqBrQsW2XrU:b9yd4o8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_primary_key.go#L232)):*
> ```Go
> 	var err error
> 	if localityConfigSwap != nil {
> 		switch localityConfigSwap.OldLocalityConfig.Locality.(type) {
> ```

Is there any way we could build helper functions to make this code read easier?  For example:

```
ol := localityConfigSwap.GetOldLocalityConfig()
if ol.IsLocalityGlobal() || ol.IsLocalityRegionalByRow() {
   nl := localityConfigSwap.GetNewLocalityConfig()
   if nl.IsLocalityRegionalByRow() {
      <make the magic happen>
   } else {
      assert
   }
}  else {
   assert
}
```
___
*[pkg/sql/alter_primary_key.go, line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSXyQ-T2XK51_UhT-eL:-MSXyQ-T2XK51_UhT-eM:b-xyjiop) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_primary_key.go#L321)):*
> ```Go
> 				col, err := tableDesc.FindColumnByID(colID)
> 				if err != nil {
> 					panic(err)
> ```

More of a question for @ajwerner or @lucy-zhang but is there a reason why we're panicking here, instead of, say, building an AssertionFailed error and returning it up the stack?  
___
*[pkg/sql/alter_primary_key.go, line 369 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY--Sn1NIiDZUwSVg_:-MSY--Sn1NIiDZUwSVga:bz80h1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_primary_key.go#L369)):*
> ```Go
> 			}
> 		}
> 		// TODO(#multiregion): decide what happens if any REGIONAL BY ROW/GLOBAL/
> ```

Nit: Do we need an issue number here?
___
*[pkg/sql/alter_primary_key.go, line 372 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY0a_BAWRO3q8ve68H:-MSY0a_BAWRO3q8ve68I:bs039j8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_primary_key.go#L372)):*
> ```Go
> 		//                     normal tables have PARTITION BY statements.
> 		// Detect partitioning if PARTITION BY ALL.
> 		if partitionAllBy != nil && isNewPartitionAllBy {
> ```

Is it not sufficient to check for `isNewPartitionAllBy` here?  When can `isNewPartitionAllBy` be `true` and `partitionAllBy` be `nil`?  If I'm following things correctly, `isNewPartitionAllBy` is only set to true in the same block where we generate `partitionAllBy`. Maybe an assert instead of the check?
___
*[pkg/sql/alter_table_locality.go, line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY5Gd2-2mq8_nWtQGm:-MSY5Gd2-2mq8_nWtQGn:b-9q35tk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_table_locality.go#L184)):*
> ```Go
> 	newLocality *tree.Locality,
> ) error {
> 	if newLocality.RegionalByRowColumn == "" {
> ```

Nit: can we create a constant here for `""` like we have for `tree.PrimaryRegionLocalityName`?  Maybe something like `tree.RegionalByRowUnspecifiedCol`? 
___
*[pkg/sql/alter_table_locality.go, line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY9Xkv-yr2YbctMyq6:-MSY9Xkv-yr2YbctMyq7:b-2smgrt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_table_locality.go#L222)):*
> ```Go
> 	}
> 
> 	if err := params.p.AlterPrimaryKey(
> ```

There's a lot of magic happening in this call.  Can we add a comment here which explains that this not only alters the table and all of its indexes to be partitioned, but also adds the zone configs?  Without it, people may have to do some unnecessary digging to understand this code.
___
*[pkg/sql/alter_table_locality.go, line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY7Ieh1wUneoOx-ZAn:-MSY7Ieh1wUneoOx-ZAo:b4y5j6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_table_locality.go#L242)):*
> ```Go
> 		return err
> 	}
> ```

Any benefit to validating the LocalityConfig here (like we do for the other alter patterns)?  Or is that not possible because of the asynchronous nature of the above call?
___
*[pkg/sql/schema_changer.go, line 1153 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY9sqY2nKP-5BN60IT:-MSY9sqY2nKP-5BN60IU:bkvg0uu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/schema_changer.go#L1153)):*
> ```Go
> 				if lcSwap := pkSwap.LocalityConfigSwap; lcSwap != nil {
> 					// For locality configs, we need to update the zone configs immediately.
> 					switch lcSwap.NewLocalityConfig.Locality.(type) {
> ```

Nit: Could use new helper functions here too, if you created them above.
___
*[pkg/sql/schema_changer.go, line 1252 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSYAIRZ6kjJ_qQO_c0b:-MSYAIRZ6kjJ_qQO_c0c:b-mc2utc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/schema_changer.go#L1252)):*
> ```Go
> 				if lcSwap := pkSwap.LocalityConfigSwap; lcSwap != nil {
> 					scTable.LocalityConfig = lcSwap.NewLocalityConfig
> 					switch lcSwap.NewLocalityConfig.Locality.(type) {
> ```

Nit: Validate descriptors LocalityConfig down here?
___
*[pkg/sql/catalog/descpb/structured.proto, line 566 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSXtuNXCq-0o1rdOrrR:-MSXtuNY0Jrx3xRLqhIJ:bwjaeo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/catalog/descpb/structured.proto#L566)):*
> ```protobuf
>   message LocalityConfigSwap {
>     option (gogoproto.equal) = true;
>     optional TableDescriptor.LocalityConfig old_locality_config = 1;
> ```

Should these both be optional?  I'd think that if the optional LocalityConfigSwap is defined, both of these would need to be set.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59624)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @ajwerner, @arulajmani, and @lucy-zhang)

___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 723 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSXqSRh4kM9Sz1U4Qxe:-MSZDXLH9QU2VF-g4ILp:bp9j3v1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L723)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Nit: This test was meant to test going to RBR (from RBT <no region>) without the column there.  The new test you've added is great, but I want to make sure that we don't lose the old test in the process.  Can you add back an unimplemented test where we first try to move it to RBR without specifying the AS column?
</blockquote></details>

it's there. see below.

```
statement error unimplemented: implementation pending
ALTER TABLE regional_by_table_no_region SET LOCALITY REGIONAL BY ROW
```
___
*[pkg/ccl/partitionccl/regional_by_row_test.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSYApGO4xfpBjEFK0eg:-MSZDc_p7vUa2-GLu8jo:b3atqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/298e651a6a1318c7a3727e88b7d8961a7f06fb0e/pkg/ccl/partitionccl/regional_by_row_test.go#L56)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Nit: checking if I'm paying attention?  :joy: 
</blockquote></details>

⛈️
___
*[pkg/sql/alter_primary_key.go, line 232 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSXx87f9l_jy8XTOq6N:-MS_5LWK79h7y0LJqKgZ:bbr7d1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_primary_key.go#L232)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Is there any way we could build helper functions to make this code read easier?  For example:

```
ol := localityConfigSwap.GetOldLocalityConfig()
if ol.IsLocalityGlobal() || ol.IsLocalityRegionalByRow() {
   nl := localityConfigSwap.GetNewLocalityConfig()
   if nl.IsLocalityRegionalByRow() {
      <make the magic happen>
   } else {
      assert
   }
}  else {
   assert
}
```
</blockquote></details>

i can make `make the magic happen` cleaner, but i think with some changes still to be added in this section as well as local variables that need to come along later, not sure it helps too much. how strongly do you feel about this?
___
*[pkg/sql/alter_primary_key.go, line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSXyQ-T2XK51_UhT-eL:-MS_49180SzEzIGI6YNc:b-m4nkz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_primary_key.go#L321)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

More of a question for @ajwerner or @lucy-zhang but is there a reason why we're panicking here, instead of, say, building an AssertionFailed error and returning it up the stack?  
</blockquote></details>

I've fixed this one up.
___
*[pkg/sql/alter_primary_key.go, line 372 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY0a_BAWRO3q8ve68H:-MSZMufK64tIIwK7yIix:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_primary_key.go#L372)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Is it not sufficient to check for `isNewPartitionAllBy` here?  When can `isNewPartitionAllBy` be `true` and `partitionAllBy` be `nil`?  If I'm following things correctly, `isNewPartitionAllBy` is only set to true in the same block where we generate `partitionAllBy`. Maybe an assert instead of the check?
</blockquote></details>

Done.
___
*[pkg/sql/alter_table_locality.go, line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY9Xkv-yr2YbctMyq6:-MSZNmVv-NglF3iJ8sDD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_table_locality.go#L222)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

There's a lot of magic happening in this call.  Can we add a comment here which explains that this not only alters the table and all of its indexes to be partitioned, but also adds the zone configs?  Without it, people may have to do some unnecessary digging to understand this code.
</blockquote></details>

Done.
___
*[pkg/sql/alter_table_locality.go, line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY7Ieh1wUneoOx-ZAn:-MSZNhJMBmTmmtEtxp0e:b-deygtn) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_table_locality.go#L242)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Any benefit to validating the LocalityConfig here (like we do for the other alter patterns)?  Or is that not possible because of the asynchronous nature of the above call?
</blockquote></details>

Nah, should be done at the schema changer level.
___
*[pkg/sql/schema_changer.go, line 1153 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY9sqY2nKP-5BN60IT:-MS_67NHAAbPoEJWTBLZ:bxo8p5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/schema_changer.go#L1153)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Nit: Could use new helper functions here too, if you created them above.
</blockquote></details>

i gave this a go and it's basically the same amount of code because the bottleneck is all the arguments you need to pass in and didn't necessarily feel cleaner :P i'm not sure it's worth it.


<!-- Sent from Reviewable.io -->

ajstorm reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59624)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @ajwerner, @arulajmani, @lucy-zhang, and @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 723 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSXqSRh4kM9Sz1U4Qxe:-MS_A1i6CvokMbM58XCS:b-4ct44q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L723)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

it's there. see below.

```
statement error unimplemented: implementation pending
ALTER TABLE regional_by_table_no_region SET LOCALITY REGIONAL BY ROW
```
</blockquote></details>

Not exactly (or at least, not in a fool-proof way).  When someone implements the test you mentioned, and does nothing else, the next test will be an alter from RBR to RBR AS.  We need another alter in between to ensure that the table ends up back in the RBT <no region> state before we execute that next test.  Can we put in that ALTER (which will also be unimplemented) now so that we don't lose track of it?
___
*[pkg/sql/alter_primary_key.go, line 232 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSXx87f9l_jy8XTOq6N:-MS_B3RxBSZhoyVpWsNs:bb6ifb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_primary_key.go#L232)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i can make `make the magic happen` cleaner, but i think with some changes still to be added in this section as well as local variables that need to come along later, not sure it helps too much. how strongly do you feel about this?
</blockquote></details>

3/10
___
*[pkg/sql/alter_primary_key.go, line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSXyQ-T2XK51_UhT-eL:-MS_BPKx1EA5KodZS0-J:b-sipalp) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_primary_key.go#L321)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I've fixed this one up.
</blockquote></details>

bueno!
___
*[pkg/sql/alter_table_locality.go, line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY9Xkv-yr2YbctMyq6:-MS_C0CT6hkoP1pnd2rp:b-8yf7ca) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_table_locality.go#L222)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Much better!  For bonus points, can you also mention that the re-partitioning also writes the zone configs?
___
*[pkg/sql/alter_table_locality.go, line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY7Ieh1wUneoOx-ZAn:-MS_D2mT1D7Mu1mOM2Gs:braelq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_table_locality.go#L242)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Nah, should be done at the schema changer level.
</blockquote></details>

Are we doing it down there?  I didn't catch it.
___
*[pkg/sql/schema_changer.go, line 1153 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY9sqY2nKP-5BN60IT:-MS_CH4Q5UelE-VfqbEV:b-1p679u) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/schema_changer.go#L1153)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i gave this a go and it's basically the same amount of code because the bottleneck is all the arguments you need to pass in and didn't necessarily feel cleaner :P i'm not sure it's worth it.
</blockquote></details>

I trust you on this one.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59624)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @ajwerner, @arulajmani, and @lucy-zhang)

___
*[pkg/sql/alter_table_locality.go, line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY7Ieh1wUneoOx-ZAn:-MS_bvGi1kHr2acNXpNO:bg5xae1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_table_locality.go#L242)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Are we doing it down there?  I didn't catch it.
</blockquote></details>

see far far in schema_changer.go:

```
					// Validate the new locality before updating the table descriptor.
					dg := catalogkv.NewOneLevelUncachedDescGetter(txn, sc.execCfg.Codec)
					if err := scTable.ValidateTableLocalityConfig(
						ctx,
						dg,
					); err != nil {
						return err
					}

```


<!-- Sent from Reviewable.io -->

ajstorm reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59624)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @ajwerner, @arulajmani, and @lucy-zhang)

___
*[pkg/sql/alter_table_locality.go, line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY7Ieh1wUneoOx-ZAn:-MS_dwbv7rXetH3snpKo:bjwa7hh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_table_locality.go#L242)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

see far far in schema_changer.go:

```
					// Validate the new locality before updating the table descriptor.
					dg := catalogkv.NewOneLevelUncachedDescGetter(txn, sc.execCfg.Codec)
					if err := scTable.ValidateTableLocalityConfig(
						ctx,
						dg,
					); err != nil {
						return err
					}

```
</blockquote></details>

:tada: 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59624)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @lucy-zhang)

___
*[pkg/sql/alter_primary_key.go, line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSXyQ-T2XK51_UhT-eL:-MSYDJEc20adnENbm1Bc:by4agwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_primary_key.go#L321)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

More of a question for @ajwerner or @lucy-zhang but is there a reason why we're panicking here, instead of, say, building an AssertionFailed error and returning it up the stack?  
</blockquote></details>

I think it's mostly that it's in a closure that doesn't have an error return. The idea is just that this shouldn't happen by construction as, in theory, there was some validation that happened earlier to validate that the indexes were well formed. I'm cool with the idea of adding an error return. 


<!-- Sent from Reviewable.io -->

arulajmani reviewed (changes requested):
First pass! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-:-MSclZUk28XAYdHllbcg:btsrnr9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @ajwerner, @lucy-zhang, and @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScPbwt0krnpb0q2Si4:-MScPbwt0krnpb0q2Si5:bae63l3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L72)):*
> ```
> statement ok
> CREATE TABLE regional_by_table_in_primary_region (
>   pk INT PRIMARY KEY,
> ```

Why the change (here and other places)?
___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 401 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScQHKx7sDHy5V5gMyR:-MScQHKx7sDHy5V5gMyS:b-ibzghf) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L401)):*
> ```
> ) LOCALITY REGIONAL BY ROW AS cr;
> ALTER TABLE alter_locality_test.public.created_as_global CONFIGURE ZONE USING
>   global_reads = true;
> ```

Should global_reads be true for a regional by row table here?
___
*[pkg/sql/alter_primary_key.go, line 263 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScSvLF95ZREd_pR5NN:-MScSvLF95ZREd_pR5NO:btp0t3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_primary_key.go#L263)):*
> ```Go
> 			}
> 		default:
> 			return errors.AssertionFailedf(
> ```

Do we not need a case for when the "old" locality config is regional by row? 
___
*[pkg/sql/alter_primary_key.go, line 369 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY--Sn1NIiDZUwSVg_:-MScTfXZ3bqzNObBdZfj:b6yji4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_primary_key.go#L369)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Nit: Do we need an issue number here?
</blockquote></details>

nit:  TODO(otan) / TODO(multiregion) with issue number in the comment.

Also, the issue description and the comment here seem to diverge. The issue description reads "sql: decide whether we allow PARTITION [ALL] BY for non-REGIONAL BY ROW tables", where as this comment seems to indicate that `PARTITION [ALL] BY` may be something that all tables have and we have to deal with it here. Which one is it?
___
*[pkg/sql/alter_primary_key.go, line 381 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScTk547edI4_uA9Lzh:-MScTk547edI4_uA9Lzi:b-2spzg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/77d03d1833435192b44f38f378e93b132f24fbd6/pkg/sql/alter_primary_key.go#L381)):*
> ```Go
> 		}
> 		// TODO(#59719): decide what happens if any REGIONAL BY ROW/GLOBAL/
> 		//               normal tables have PARTITION BY statements.
> ```

nit: extra spaces. 
___
*[pkg/sql/alter_primary_key.go, line 387 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScj4ZyA7_jQq8jfKYv:-MScj4ZyA7_jQq8jfKYw:bbzb7u5) ([raw file](https://github.com/cockroachdb/cockroach/blob/77d03d1833435192b44f38f378e93b132f24fbd6/pkg/sql/alter_primary_key.go#L387)):*
> ```Go
> 		// will make these indexes explicit.
> 		if isNewPartitionAllBy {
> 			if *newIndex, err = CreatePartitioning(
> ```

This only partitions the primary index, right? How do secondary indexes get partitioned?
___
*[pkg/sql/alter_table_locality.go, line 238 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSckzVcCqRJ-yKKzS0a:-MSckzVcCqRJ-yKKzS0b:b-cxueuo) ([raw file](https://github.com/cockroachdb/cockroach/blob/77d03d1833435192b44f38f378e93b132f24fbd6/pkg/sql/alter_table_locality.go#L238)):*
> ```Go
> 			Columns: cols,
> 		},
> 		&descpb.PrimaryKeySwap_LocalityConfigSwap{
> ```

You could pull this proto construction out in a method. 
___
*[pkg/sql/region_util.go, line 718 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSceFd030n6zpllhH1A:-MSceFd030n6zpllhH1B:bqu7bet) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/region_util.go#L718)):*
> ```Go
> 	}
> 
> 	return &tree.PartitionBy{
> ```

Im probably missing something, but why is this not constructing a `PartitionByTable` node with `All` set to true? Isn't that what we want in the case of regional by row tables? 
___
*[pkg/sql/schema_changer.go, line 1162 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScdNlF0E0_5QcnCxaD:-MScdNlG57uwzG7eQpw4:b6wc56k) ([raw file](https://github.com/cockroachdb/cockroach/blob/77d03d1833435192b44f38f378e93b132f24fbd6/pkg/sql/schema_changer.go#L1162)):*
> ```Go
> 					}
> 
> 					// For locality configs, we need to update the zone configs immediately.
> ```

Could use some more context/reword. As is, I'm left wondering why we need to update them immediately and why this behavior differs from when this code path is hit in an alter primary key change. 
___
*[pkg/sql/catalog/descpb/structured.proto, line 569 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScbsItAk3AINMin95b:-MScbsItAk3AINMin95c:bkkubs1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/catalog/descpb/structured.proto#L569)):*
> ```protobuf
>     optional TableDescriptor.LocalityConfig new_locality_config = 2;
>   }
>   // LocalityConfigSwap is set for ALTER TABLE SET LOCALITY when the command requires PK changes.
> ```

nit: wrap comment to 80 chars please.
___
*[pkg/sql/sem/tree/region.go, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSccHJH46rxi243ZQDe:-MSccHJH46rxi243ZQDf:b-o7ogka) ([raw file](https://github.com/cockroachdb/cockroach/blob/5241ff01d1a073861e47ef2cd003666614654f4c/pkg/sql/sem/tree/region.go#L41)):*
> ```Go
> 	// RegionalByRowRegionNotSpecifiedName is the string denoting the REGIONAL BY ROW
> 	// has no AS <col> specified.
> 	RegionalByRowRegionNotSpecifiedName = ""
> ```

Can we hijack this in the parser itself and replace it with `RegionalByRowRegionDefaultColumnName`? I'm not a big fan of this empty string meaning default column in a bunch of places. To be fair, this was an issue before your change as well, so feel free to disregard if you consider it orthogonal. 


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59624)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @lucy-zhang)

___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScPbwt0krnpb0q2Si4:-MSdItZg0ogMpZ0CT8kf:b-ysqkbn) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L72)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Why the change (here and other places)?
</blockquote></details>

Because logictests randomises families, and rowid can be in random family, making SHOW CREATE TABLE flaky. being specific here makes the randomisation issue go away.
___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 401 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScQHKx7sDHy5V5gMyR:-MSdIr9yCz10HbCjUzNT:bw0xf6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L401)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should global_reads be true for a regional by row table here?
</blockquote></details>

You have been reading an old revision.
___
*[pkg/sql/alter_primary_key.go, line 263 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScSvLF95ZREd_pR5NN:-MSdJ5JS3sYnd87Zf-BV:b-2hp2mx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_primary_key.go#L263)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Do we not need a case for when the "old" locality config is regional by row? 
</blockquote></details>

Yes, but we haven't supported it yet.
so hence, unsupported.
___
*[pkg/sql/alter_primary_key.go, line 369 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSY--Sn1NIiDZUwSVg_:-MSdJIbE7sQHjLafv8RL:b-mo6u5n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_primary_key.go#L369)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit:  TODO(otan) / TODO(multiregion) with issue number in the comment.

Also, the issue description and the comment here seem to diverge. The issue description reads "sql: decide whether we allow PARTITION [ALL] BY for non-REGIONAL BY ROW tables", where as this comment seems to indicate that `PARTITION [ALL] BY` may be something that all tables have and we have to deal with it here. Which one is it?
</blockquote></details>

more a fan of TODO(#issue) than TODO(#owner) in this case. owners can change but issue numbers are 4eva.

i've fixed both issue title and here to match.
___
*[pkg/sql/alter_primary_key.go, line 387 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScj4ZyA7_jQq8jfKYv:-MSdJV_Y4Gg8w0iiQE06:bxqiy9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/77d03d1833435192b44f38f378e93b132f24fbd6/pkg/sql/alter_primary_key.go#L387)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This only partitions the primary index, right? How do secondary indexes get partitioned?
</blockquote></details>

No, this bit partitions the secondary indexes. Primary index partitioning is done above.
___
*[pkg/sql/alter_table_locality.go, line 238 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSckzVcCqRJ-yKKzS0a:-MSdMrMK40T8W33Qk6vn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/77d03d1833435192b44f38f378e93b132f24fbd6/pkg/sql/alter_table_locality.go#L238)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

You could pull this proto construction out in a method. 
</blockquote></details>

Done.
___
*[pkg/sql/region_util.go, line 718 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSceFd030n6zpllhH1A:-MSdKXhj4kt-2MeIYLGU:b-fnu79i) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/region_util.go#L718)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Im probably missing something, but why is this not constructing a `PartitionByTable` node with `All` set to true? Isn't that what we want in the case of regional by row tables? 
</blockquote></details>

We don't need to entire PartitionByTable clause for anything, so doesn't feel like there's a need.
___
*[pkg/sql/schema_changer.go, line 1162 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScdNlF0E0_5QcnCxaD:-MSdLHfd-bxFdN_MfOLF:b4t9thd) ([raw file](https://github.com/cockroachdb/cockroach/blob/77d03d1833435192b44f38f378e93b132f24fbd6/pkg/sql/schema_changer.go#L1162)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Could use some more context/reword. As is, I'm left wondering why we need to update them immediately and why this behavior differs from when this code path is hit in an alter primary key change. 
</blockquote></details>

The context is provided above this green blob. I've mentioned why we overwrite instead of copy, however.

```
				// We might have to update some zone configs for indexes that are
				// being rewritten. It is important that this is done _before_ the
				// index swap occurs. The logic that generates spans for subzone
				// configurations removes spans for indexes in the dropping state,
				// which we don't want. So, set up the zone configs before we swap.
```
___
*[pkg/sql/sem/tree/region.go, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSccHJH46rxi243ZQDe:-MSdKxpB1_TH2pN2uOgB:bn6kxg1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5241ff01d1a073861e47ef2cd003666614654f4c/pkg/sql/sem/tree/region.go#L41)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Can we hijack this in the parser itself and replace it with `RegionalByRowRegionDefaultColumnName`? I'm not a big fan of this empty string meaning default column in a bunch of places. To be fair, this was an issue before your change as well, so feel free to disregard if you consider it orthogonal. 
</blockquote></details>

i've copy pasted your TODO.
Hey look, TODO(#issuenum) :P 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just the questions about behavior as this change pertains to already partitioned tables and secondary indexes which already have zone configurations. 

> Reviewed 2 of 17 files at r3, 7 of 14 files at r5, 6 of 7 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-:-MSdcwmP7eND9g6hy-Dm:b-t7m5xm)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, @lucy-zhang, and @otan)

___
*[pkg/ccl/partitionccl/regional_by_row_test.go, line 152 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSdbYR8BEKAOcp-_eFM:-MSdbYR8BEKAOcp-_eFN:b-zfjq6z) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ba8133eaca7fe2d6545494fa0720417788832d7/pkg/ccl/partitionccl/regional_by_row_test.go#L152)):*
> ```
> 	// Stop any further attempts at cancellation, so the GC jobs don't fail.
> 	shouldCancel = false
> ```
Why would the GC job fail? It doesn't use backfills, does it? 
___
*[pkg/sql/alter_primary_key.go, line 381 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSdcBA5A-u9K0S1It8H:-MSdcBA5A-u9K0S1It8I:b2dm3mu) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ba8133eaca7fe2d6545494fa0720417788832d7/pkg/sql/alter_primary_key.go#L381)):*
> ```
> 		// TODO(#59719): decide what happens if any multiregion tables have
> 		// PARTITION BY statements.
> ```
Should we put some gating here? Is it just going to throw away the partitioning now or is that handled somewhere I'm not seeing?
___
*[pkg/sql/region_util.go, line 467 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSdad-b18XN7RTa7-A6:-MSdad-b18XN7RTa7-A7:bulotoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/77d03d1833435192b44f38f378e93b132f24fbd6/pkg/sql/region_util.go#L467)):*
> ```Go
> // applyZoneConfigForMultiRegionTableOptionTable applies table zone configs
> // on the entire table with the new locality config.
> func applyZoneConfigForMultiRegionTableOptionTableNewConfig(
> ```

nit:  `applyZoneConfigForMultiRegionTableOptionTable` is both very long and not the name of this function.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59624)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm, @ajwerner, @arulajmani, and @lucy-zhang)

___
*[pkg/ccl/partitionccl/regional_by_row_test.go, line 152 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSdbYR8BEKAOcp-_eFM:-MSeZrj48cykSo4dR7_2:b-h5apmy) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ba8133eaca7fe2d6545494fa0720417788832d7/pkg/ccl/partitionccl/regional_by_row_test.go#L152)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	// Stop any further attempts at cancellation, so the GC jobs don't fail.
> 	shouldCancel = false
> ```
Why would the GC job fail? It doesn't use backfills, does it? 
</blockquote></details>

Copy pasta! removed.
___
*[pkg/sql/alter_primary_key.go, line 381 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSdcBA5A-u9K0S1It8H:-MSe_-o25HqFd-H7RW18:be5dk5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ba8133eaca7fe2d6545494fa0720417788832d7/pkg/sql/alter_primary_key.go#L381)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 		// TODO(#59719): decide what happens if any multiregion tables have
> 		// PARTITION BY statements.
> ```
Should we put some gating here? Is it just going to throw away the partitioning now or is that handled somewhere I'm not seeing?
</blockquote></details>

We're throwing it away for now, which is ok. We can come back to it if we decide to do something with them (which is unlikely)
___
*[pkg/sql/region_util.go, line 467 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSdad-b18XN7RTa7-A6:-MSepmrX-q5YGECUeeC1:bvhqm3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/77d03d1833435192b44f38f378e93b132f24fbd6/pkg/sql/region_util.go#L467)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit:  `applyZoneConfigForMultiRegionTableOptionTable` is both very long and not the name of this function.
</blockquote></details>

hmm i think i make this less verbose when these functions move to CCL / and we can remove "MultiRegion" for the name (the sql package needs some disambiguation to avoid confusion which is why they've been done this way). i've fixed the comment up for now - if there's a better shorter name i'm happy to do so, but may hold off on it because there's other things with the same naming scheme.


<!-- Sent from Reviewable.io -->

otan committed:
sql: add test for ALTER TABLE ... LOCALITY REGIONAL BY ROW / CANCEL

Since LOCALITY REGIONAL BY ROW requires CCL, we have to move some test
utilities to sqltestutils and test from a CCL package. This commit
covers the relevant moves as well as adding the new test.

Release note: None

ajwerner reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-:-MSfLcc_4TIx3wQwqBqw:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm, @ajwerner, @arulajmani, and @lucy-zhang)

___
*[pkg/sql/region_util.go, line 467 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSdad-b18XN7RTa7-A6:-MSfLRKvDPQBaedTLXq7:bqrfan2) ([raw file](https://github.com/cockroachdb/cockroach/blob/77d03d1833435192b44f38f378e93b132f24fbd6/pkg/sql/region_util.go#L467)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

hmm i think i make this less verbose when these functions move to CCL / and we can remove "MultiRegion" for the name (the sql package needs some disambiguation to avoid confusion which is why they've been done this way). i've fixed the comment up for now - if there's a better shorter name i'm happy to do so, but may hold off on it because there's other things with the same naming scheme.
</blockquote></details>

Understood.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
One suggestion, but otherwise <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-:-MShWXJN-2vUpnM64Ty7:b-e8iuwe)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @ajstorm, @ajwerner, @arulajmani, @lucy-zhang, and @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 401 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScQHKx7sDHy5V5gMyR:-MShQsE3AgMxjWsSWgmB:b-7g5864) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L401)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

You have been reading an old revision.
</blockquote></details>

yikes
___
*[pkg/sql/alter_primary_key.go, line 263 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScSvLF95ZREd_pR5NN:-MShREYU7Wql-GTuwX_m:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/alter_primary_key.go#L263)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Yes, but we haven't supported it yet.
so hence, unsupported.
</blockquote></details>

:+1: 
___
*[pkg/sql/alter_primary_key.go, line 387 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScj4ZyA7_jQq8jfKYv:-MShWLm27wH8HsCZK-6_:b-b7e48e) ([raw file](https://github.com/cockroachdb/cockroach/blob/77d03d1833435192b44f38f378e93b132f24fbd6/pkg/sql/alter_primary_key.go#L387)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

No, this bit partitions the secondary indexes. Primary index partitioning is done above.
</blockquote></details>

Oops I thought this was outside the for loop when I left the comment. nvm me :) 
___
*[pkg/sql/region_util.go, line 718 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSceFd030n6zpllhH1A:-MShTgVK7BctzFLjgCbU:b3jthgx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/region_util.go#L718)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

We don't need to entire PartitionByTable clause for anything, so doesn't feel like there's a need.
</blockquote></details>

Looks like a couple of places where this method is called from either set `PartitionAllBy` on the descriptor or store that information in a variable. If you return an explicit `PartitionByTable` node with `All` set to true here, it'll localize this relationship to this method as opposed to having the implicit relationship being established in a few places. Not something to hold this PR up, but I'd really like if we made this change before merging (if you agree, that is).  
___
*[pkg/sql/schema_changer.go, line 1162 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MScdNlF0E0_5QcnCxaD:-MShRf8kDioCkvIQb-3Z:b-h1pd30) ([raw file](https://github.com/cockroachdb/cockroach/blob/77d03d1833435192b44f38f378e93b132f24fbd6/pkg/sql/schema_changer.go#L1162)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

The context is provided above this green blob. I've mentioned why we overwrite instead of copy, however.

```
				// We might have to update some zone configs for indexes that are
				// being rewritten. It is important that this is done _before_ the
				// index swap occurs. The logic that generates spans for subzone
				// configurations removes spans for indexes in the dropping state,
				// which we don't want. So, set up the zone configs before we swap.
```
</blockquote></details>

Thanks, this reads much nicer! 
___
*[pkg/sql/sem/tree/region.go, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSccHJH46rxi243ZQDe:-MShSK6CDu-Hc546cQgc:bjdtd2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/5241ff01d1a073861e47ef2cd003666614654f4c/pkg/sql/sem/tree/region.go#L41)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i've copy pasted your TODO.
Hey look, TODO(#issuenum) :P 
</blockquote></details>

To be fair, you're the one who did the TODO(#issuenum) change on my original TODO in the first place -.-" 

I'm not against this in principal, just that it diverges from all other TODOs we have in the codebase. 


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59624)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @ajstorm, @ajwerner, @arulajmani, @lucy-zhang, and @otan)

___
*[pkg/sql/region_util.go, line 718 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSceFd030n6zpllhH1A:-MSiTX3wEFiiqFKSaJXw:bwkb1do) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddfa124c4b712ba3d44486c6a9b99370df7d750/pkg/sql/region_util.go#L718)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Looks like a couple of places where this method is called from either set `PartitionAllBy` on the descriptor or store that information in a variable. If you return an explicit `PartitionByTable` node with `All` set to true here, it'll localize this relationship to this method as opposed to having the implicit relationship being established in a few places. Not something to hold this PR up, but I'd really like if we made this change before merging (if you agree, that is).  
</blockquote></details>

not quite; we have isNewPartitionAllBy (this distinction is important) for one, and we need to store it directly on the descriptor for the other usage.
i don't see this as being a huge advantage without more code refactors right now, but happy for you to prove me wrong in a separate commit ;D 
___
*[pkg/sql/sem/tree/region.go, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59624#-MSccHJH46rxi243ZQDe:-MSiSu0l17B79jKlWLo4:b2vv9zj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5241ff01d1a073861e47ef2cd003666614654f4c/pkg/sql/sem/tree/region.go#L41)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

To be fair, you're the one who did the TODO(#issuenum) change on my original TODO in the first place -.-" 

I'm not against this in principal, just that it diverges from all other TODOs we have in the codebase. 
</blockquote></details>

Hahahahaha oops :D 
As long as there's something to point it to (and not TODO(rohany))


<!-- Sent from Reviewable.io -->

otan commented:
bors r=arulajmani,ajwerner
